File: week-5/19-RemoveNthNodeFromEndOfList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if not head.next:
            return None
        dummy = ListNode()
        dummy.next = head
        slow, fast = dummy, head
        for _ in range(n):
            if fast:
                fast = fast.next
        while fast:
            slow, fast = slow.next, fast.next
        slow.next = slow.next.next
        return dummy.next

File: week-5/19-RemoveNthNodeFromEndOfList.java

public class RemoveNthNode {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head.next == null)
            return null;
        ListNode dummyNode = new ListNode();
        dummyNode.next = head;
        ListNode slow = dummyNode;
        ListNode fast = head;
        for (int i = 0; i < n; i++) {
            if(fast != null){
                fast = fast.next;
            }
        }
        while (fast != null){
            slow = slow.next;
            fast = fast.next;
        }
        slow.next = slow.next.next;
        return dummyNode.next;

    }
}
